### PR TITLE
add VNCReader and related classes and utilities

### PR DESCRIPTION
**Pull request by jonathathan**

### CODE REVIEW

**Comment by jonathathan** [comment]
`universe/readers/*` are moved from `gym-demonstrations` (lightly touched up to remove some dependencies)

`example/readers/vnc_reader.py` is moved from `gym-demonstrations`

`example/readers/screenshots.py` is new

**Comment by jonathathan** [comment]
@gdb @nottombrown moved some VNCReader stuff to `universe/readers/*` and `example/readers/*` -- locations OK?

**Comment by nottombrown** [comment]
Hey! I'm forgetting, what's the reason that we want to move this to Universe?

A couple things to think about:

1. This separates the reader from its tests at https://github.com/openai/gym-demonstration/blob/master/tests/test_event_readers.py
2. There is currently a long-outstanding branch at `gym-demonstration:0-2-dev` that may conflict with this and needs to be merged (currently slated to merge on weds)
3. We should likely have a sibling PR to gym-demonstration that removes this code so that we don't have duplicates lying around

**Comment by jonathathan** [comment]
@nottombrown the time dilation test harness uses `VNCReader`, and I think other code outside of `gym-demonstrations` could make use of it. `vnc_reader.py` is also just really useful, and since we've open-sourced the code that writes the fbs files, it seems nice to open-source the code that reads them :) I could make the test harness depend on `gym-demonstrations`, but it seems cleaner (and potentially nicer to the community) to promote the chunk that it uses.

1. Didn't notice the tests -- I'll move those in too.
2/3. I intend to pull the code out of `gym-demonstrations`. I was waiting until the dev branch was merged just to make life a little easier (it looks easy -- just changing a few imports).

**Comment by liber145** [comment]
Hi, @jonathathan. I would like to collect images and actions (key arrows) when playing games by humans. Then I could initialize my model by human performance, like the [blog](https://openai.com/blog/universe/) said. In your example `vnc_recorder.py`, it shows that I can read `client.fbs` which may store what I need. But I don't know how to get this file. 

I think maybe it needs to run `vnc_recorder.py` first, but I did not know how to get that example work either. :( Could you help me about it?

**Comment by jonathathan** [comment]
Hey @liber145, the VNC server and rewarder server listen on 5900 and 15900 respectively, but there are proxy servers listening on 5899 and 15899 that record all network traffic to disk (e.g. creating the `server.fbs` and `client.fbs` files)

You can connect to these servers manually with a VNC client (OSX ships with TigerVNC, although I recommend TurboVNC for the extra encoding options), play the game while recording your actions and observations, and then fetch the files from `/tmp/demo/` and process them using the code in this PR. 

i.e. you should be able to run 

```docker run -ti --cap-add NET_ADMIN --privileged --cap-add SYS_PTRACE --ipc host -p 5900:5900 -p 5899:5899 -p 15900:15900 -p 15899:15899 quay.io/openai/universe.flashgames run -e flashgames.NeonRace-v0``` 

which forwards recording and non-recording ports, then run `open vnc://localhost:5899` (or connect manually with whatever VNC software you have). After closing the VNC session, you should see a directory inside the container's `/tmp/demo` that contains `server.fbs` and `client.fbs`.

Complete tools for recording, processing, and learning from demos are in the works, but if you're looking to try it out before it's merged in, you can get started with the `VNCReader` class in this PR. Let me know how it goes, happy to provide support!

**Comment by liber145** [comment]
Hey, @jonathathan. Thanks for your help. I have followed tlb's steps to collect `server.fbs` and `client.fbs`, and noticed your messages just now. And you shows a way how to read these binary files. Ah, that's exactly what I want to do now. I will get started with `VNCReader` class. Thank you very much! :)

**Comment by liber145** [comment]
Hey, @jonathathan. I can read `server.fbs` and `client.fbs` with the help of `vnc_reader.py` (add flag `-v /tmp/demo:/tmp/demo` to your script, then I can collect these two files in local `/tmp/demo` rather than in docker filesystem.). The following is based on the example code in `reader`.
```python
import logging
import sys
from universe.readers.vnc_reader import VNCMultiReader

logger = logging.getLogger()

def main():
    reader = VNCMultiReader(["/tmp/demo/1486473746-kgeyzilxs96fmg-1/"],
            metadata_file=None,
            reward_file=None,
            observation_file='server.fbs',
            action_file='client.fbs',
            botaction_file=None,
            fps=5,
            crop_to_episode=False,
            paint_cursor=False,
            disable_sticky_done_signal=False
            )

    for i, (observation, reward, done, info, action) in enumerate(reader):
        if observation is None:
            continue

        print('observation shape: {}'.format(observation.shape))
        if action != []:
            print('action: {}'.format(action))
            logger.info('Action: {}'.format(action))
        if reward != 0.:
            print('reward: {}'.format(reward))
            logger.info('Reward: {}'.format(reward))
        if done:
            print('Done: {}'.format(done))
            logger.info('Done: {}'.format(done))
        if info != []:
            print('info: {}'.format(info))
            logger.debug('Info: {}'.format(info))

if __name__ == '__main__':
    sys.exit(main())
```

There are three files `reward_file`, `metadata_file`, `botaction_file` that I did not get. Among them, `reward_file` can give reward information which is needed when I learn reinforcement learning. I think I may miss some operations when listening on rewarder server port. Could you show me how to collect that file?

File: universe/readers/event_readers.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+
+from universe import error
+from universe import spaces
+from universe import utils
+from universe.vncdriver import fbs_reader, vnc_proxy_server, vnc_client
+from universe import spaces as vnc_spaces
+
+logger = logging.getLogger(__name__)
+
+class InvalidEventLogfileError(error.Error):
```
Comment (seman) [inline universe/readers/event_readers.py]: Add another blank line. There are a lot of lint error, consider running flake8 before pushing the code.

File: universe/readers/event_readers.py [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+        self.pixel_format.append(server_pixel_format)
+
+    def pop_all(self):
+        output = self.actions, self.pixel_format
+        self.actions = []
+        self.pixel_format = []
+        return output
+
+
+class RewardReader(object):
```
Comment (seman) [inline universe/readers/event_readers.py]: There are a lot of common code repeated in these RewardReader, ActionReader, BotactionReader and other reader classes. Consider refactoring it. Consider using inheritance and moving the common code to the base class.

**Comment by nottombrown** [comment]
Thanks for the feedback all. We're not ready to release this code, so I'm going to close this PR until we can release everything necessary to make this useful.

**Pull request closed without merging**